contracts/chain-adapters/ForwarderBase.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.0;
+
+import { UUPSUpgradeable } from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import { EIP712CrossChainUpgradeable } from "../upgradeable/EIP712CrossChainUpgradeable.sol";
+
+abstract contract ForwarderBase is UUPSUpgradeable, EIP712CrossChainUpgradeable {
```

Comment:
How will this contract be used? I'm not sure I understand. I can't really review this contract without understanding its context

Reply:
I created it so it could act like an interface to be implemented by specific L2 forwarders (i.e. every forwarder should have the `relayTokens`/`_requireAdminSender` implemented depending on the L2/L3 format). There are also some functions which should be common in all forwarders, like the fallback function and the proxy access control, so it made sense to factor those out into another (abstract) contract.